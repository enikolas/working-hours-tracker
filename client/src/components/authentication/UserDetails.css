.user-details {
  position: absolute;
  top: 0;
  right: 20px;
  background: #5a717f;
  text-align: center;
  overflow: hidden;
  height: 29px;
  transition: all ease-in 0.5s;
}

.user-details:hover {
  height: 102px;
}

.user-details .employee-name {
  padding: 10px;
  font-weight: bold;
  color: #c5ccd1;
  font-size: 12px;
  min-width: 130px;
}

.user-details button {
  background: transparent;
  color: #fff;
  cursor: pointer;
  opacity: 0.7;
}

.user-details button:hover {
  opacity: 1;
}

.user-details button.logout {
  background-color: #909fa8;
  text-transform: lowercase;
  padding: 2px 6px;
  color: #fff;
  cursor: pointer;
  margin-top: 4px;
}

.user-details button.logout:hover {
  background-color: #c5ccd1;
}

.user-details button.logout:focus {
  outline: 1px solid #f93;
}

.user-details nav li {
  list-style: none;
}
