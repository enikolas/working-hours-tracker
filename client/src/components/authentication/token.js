const API_AUTH_TOKEN = 'achiever-auth-token';

export const setAuthToken = token => localStorage.setItem(API_AUTH_TOKEN, token);

export const getAuthToken = () => localStorage.getItem(API_AUTH_TOKEN);

export const removeAuthToken = () => localStorage.removeItem(API_AUTH_TOKEN);
