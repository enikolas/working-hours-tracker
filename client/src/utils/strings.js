const strings = {
	times: [
		{
			label: 'Entrada',
			cliKeywords: [
				'entrada', 'entrando', 'entrei',
				'chegada', 'chegando', 'cheguei',
				'arrival'
			]
		}, {
			label: 'Ida para almo√ßo',
			cliKeywords: [
				'ida', 'indo',
				'lunch', 'gone-lunch',
				'almoco', 'ida-almoco', 'indo-almoco',
				'fui-almoco', 'partiu-almoco'
			]
		}, {
			label: 'Volta do almo√ßo',
			cliKeywords: [
				'volta', 'voltando', 'voltei',
				'back', 'back-from-lunch',
				'retorno', 'retornando', 'retornei',
				'volta-almoco', 'voltando-almoco', 'voltei-almoco',
				'chegada-almoco', 'chegando-almoco', 'cheguei-almoco'
			]
		}, {
			label: 'Sa√≠da',
			cliKeywords: [
				'saida', 'saindo', 'sai',
				'fui', 'fui-casa', 'partiu-casa',
				'leave', 'left', 'gone-home',
				'ida-casa', 'indo-casa'
			]
		}
	],
	clearCliKeywords: [
		'clear', 'clean', 'reset', 'apaga'
	],
	multipleCliKeywords: [
		'multi', 'multiplos', 'varios'
	],
	sendCallFeedbacks: {
		error: 'Erro ao executar api call para inserir horas:',
		success: 'Api call para inserir horas executada com sucesso!'
	},
	thisTime: 'esta hora',
	hoursLabouredOnThisDay: 'Horas trabalhadas no dia:',
	hoursBalanceOnWeekUpToNow: 'Na semana at√© esse dia',
	hoursBalanceOnWeekUpToNowSurplus: '{0} a mais',
	hoursBalanceOnWeekUpToNowDebt: 'devendo {0}',
	hoursBalanceForToday: 'Balan√ßo de horas atual:',
	storedSuccessfully: 'gravada com sucesso',
	morningPeriod: 'Per√≠odo da manh√£',
	afternoonPeriod: 'Per√≠odo da tarde',
	weeklyCalendar: 'Calend√°rio da semana',
	total: 'Total',
	send: 'Enviar',
	update: 'Atualizar',
	ok: 'OK',
	cancel: 'Cancelar',
	yes: 'Sim',
	no: 'N√£o',
	loading: 'Carregando...',
	noTimesProvided: 'nenhuma hora foi fornecida',
	otherTimesIgnored: 'as outras horas fornecidas foram ignoradas',
	timesFlushed: 'As entradas de tempo foram apagadas',
	cannotInsertDisconnectedTime: ' n√£o pode ser inserida sem o(s) hor√°rio(s) anterior(es)',
	cannotSelectFutureTime: 'Uma data futura n√£o pode ser selecionada',
	markConfirm: 'Quer marcar agora a',
	markNow: 'Marcar agora a',
	timeSentToday: 'As horas de hoje j√° foram enviadas',
	clearConfirm: 'Quer mesmo apagar o arquivo tempor√°rio de horas pra hoje?',
	cliConfirmationAllowed: 'Digite "s" ou "n"',
	cancelled: 'Cancelado',
	todayPage: 'Horas de hoje',
	editPage: 'Editar hor√°rios',
	timeEntryPage: 'Apontamento de horas',
	selectedDate: 'Dia selecionado',
	dateBeingEdited: 'Dia sendo editado',
	todayDate: 'Dia de hoje',
	login: 'Entrada',
	logout: 'Sair',
	logoutConfirm: 'Quer mesmo sair do sistema?',
	username: 'Usu√°rio',
	password: 'Senha',
	vacation: 'F√©rias',
	authenticationError: 'Credenciais inv√°lidas.',
	pageNotFound: 'P√°gina n√£o encontrada',
	pageLoading: 'Carregando...',
	confirmSave: 'Existem altera√ß√µes n√£o salvas. Deseja enviar?',
	invalidTime: 'Hor√°rios inv√°lidos',
	invalidAddTime: 'N√£o √© poss√≠vel marcar presen√ßa em um hor√°rio anterior',
	submitTimeSuccess: 'Horas registradas com sucesso!',
	goBack: 'Voltar',
	helloName: 'Ol√° ',
	warning: 'Aten√ß√£o',
	confirmation: 'Confirma√ß√£o',
	projectPhase: 'Projeto/fase',
	activity: 'Atividade',
	noLunchPause: 'Sem pausa para almo√ßo',
	normalDay: 'Dia normal',
	hourEntryMode: 'Entrada de horas',
	specialDay: 'Dia especial',
	specialDays: {
		medical: 'Dispensa m√©dica',
		birthdayLeave: 'O-tanjoubi'
	},
	success: 'üëç',
	error: 'Erro!',
	passwordChange: 'Mudar senha',
	reportDownload: 'Relat√≥rio mensal',
	oldPassword: 'Senha antiga',
	newPassword: 'Nova senha',
	reportMonth: 'M√™s do relat√≥rio',
	selectMonth: 'selecione o m√™s...',
	insertNewPassword: 'insira a nova senha',
	confirmNewPassword: 'confirme a nova senha',
	passwordChangeSuccessful: 'A senha foi alterada com sucesso!'
};

module.exports = strings;
