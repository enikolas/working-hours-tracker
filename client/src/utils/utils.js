const _isNumber = value => !Number.isNaN(parseInt(value, 10));

export const checkValidity = (mode, valueRaw) => {
	if (valueRaw === null) {
		return true;	// 1st change on TimeGroup, will be null
	}
	if (String(valueRaw).length > 2) {
		return false;
	}

	const value = parseInt(valueRaw, 10);
	if (_isNumber(value)) {
		const max = mode === 'hours' ? 24 : 60;
		return max > value && value >= 0;
	}
	return false;
};

export const timeIsValid = time => checkValidity('minutes', time.minutes) &&
	checkValidity('hours', time.hours);

export const getTimeFromDate = (rawDate) => {
	const date = (typeof rawDate === 'number') ? new Date(rawDate) : rawDate;
	return {
		hours: date.getHours(),
		minutes: date.getMinutes()
	};
};

export const buildDateFromTimeString = (timeString) => {
	const [hours, minutes] = timeString.split(':');
	const year = new Date().getFullYear();
	const month = new Date().getMonth();
	const day = new Date().getDate();
	return new Date(year, month, day, hours, minutes);
};
