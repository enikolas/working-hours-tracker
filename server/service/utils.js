const getOptions = (cookieJar) => ({
  jar: cookieJar,
  withCredentials: true,
  rejectUnauthorized: false
});

module.exports = {
  getOptions,
};
