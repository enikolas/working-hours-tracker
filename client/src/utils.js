/* global window */
import moment from 'moment-timezone';
import TimeDuration from 'time-duration';

import strings from './utils/strings';
import { timeIsValid } from './utils/utils';

moment.locale('pt-br');
moment.tz.setDefault('America/Sao_Paulo');

export const STORAGEKEY = 'storedTimes';
export const STORAGEDAYKEY = 'storedMoment';
export const storedTimesIndex = {
	startTime: 0,
	startBreakTime: 1,
	endBreakTime: 2,
	endTime: 3
};
export const SPECIAL_ACTIVITY_HOLIDAY = { id: 99999, name: 'Holiday' };

/**
 * Check if the date1 === date2 based on day
 * @param {Date} date1
 * @param {Date} date2
 */
export const areTheSameDay = (date1, date2) => (
	date1.day() === date2.day() &&
	date1.month() === date2.month() &&
	date1.year() === date2.year()
);

/**
 * Replace the value of array[index] to be newValue
 * @param {[]} array to be replaced
 * @param {*} index of the element to be replaced
 * @param {*} newValue new value ot array[index]
 */
export const replacingValueInsideArray = (array, index, newValue) => [
	...array.slice(0, index),
	newValue,
	...array.slice(index + 1)
];

const _setStorage = (key, data) => {
	global.localStorage.setItem(key, JSON.stringify(data));
};

/**
 * This function set at storage {key, value} {dayKey: today's date}
 * @param {*} key to be used as a index key for data
 * @param {*} dayKey to be used as index key for the day
 * @param {*} data to be save as {key, data}
 */
export const setTodayStorage = (data, key = STORAGEKEY, dayKey = STORAGEDAYKEY) => {
	const today = moment();
	_setStorage(dayKey, today.valueOf());
	_setStorage(key, data);
};

const _getStorage = key => (
	JSON.parse(global.localStorage.getItem(key))
);

/**
 * Get from storage {key, data} and {dayKey, dayOnLocalStorage}. If today's date is different from
 * LocalStorage it will delete it all
 * @param {*} key is the key to be used to get {key, data}
 * @param {*} dayKey is the key to be used to get {dayKey, dayOnLocalStorage}
 */
export const getTodayStorage = (key = STORAGEKEY, dayKey = STORAGEDAYKEY) => {
	const dayOnLocal = moment(_getStorage(dayKey));
	const today = moment();
	if (global.localStorage.getItem(dayKey)) {
		if (areTheSameDay(dayOnLocal, today)) {
			return _getStorage(key);
		}
	}
	setTodayStorage({ storedTimes: [{}, {}, {}, {}], sentToday: false }, key, dayKey);
	return { storedTimes: [{}, {}, {}, {}], sentToday: false };
};

/**
 * Remove from localStorage key, dayKey
 * @param {*} key
 * @param {*} dayKey
 */
export const clearTodayStorage = (key = STORAGEKEY, dayKey = STORAGEDAYKEY) => {
	localStorage.removeItem(key);
	localStorage.removeItem(dayKey);
};

/**
 *
 * @param {Object} timeEntryInput
 * @param {Function} addTimeEntry
 */
const _addTimeEntry = async (timeEntryInput, addTimeEntry) => {
	let response;
	try {
		response = await addTimeEntry({
			variables: {
				timeEntry: timeEntryInput
			}
		});
	} catch (error) {
		return { errorMessage: error.graphQLErrors[0].message };
	}

	if (response) {
		return { successMessage: strings.submitTimeSuccess };
	}

	return { successMessage: '' };
};

/**
 *
 * @param {Date} date
 * @param {Object} stateStoredTimes
 * @param {Function} addTimeEntry
 */
export const submitToServer = async (date, stateStoredTimes, phase, activity, addTimeEntry) => {
	const startTime = stateStoredTimes[storedTimesIndex.startTime];
	const startBreakTime = stateStoredTimes[storedTimesIndex.startBreakTime];
	const endBreakTime = stateStoredTimes[storedTimesIndex.endBreakTime];
	const endTime = stateStoredTimes[storedTimesIndex.endTime];

	const timeEntryInput = {
		date: date.format('YYYY-MM-DD'),
		phaseId: phase.id,
		activityId: activity.id,
		startTime: `${startTime.hours}:${startTime.minutes}`,
		startBreakTime: `${startBreakTime.hours}:${startBreakTime.minutes}`,
		endBreakTime: `${endBreakTime.hours}:${endBreakTime.minutes}`,
		endTime: `${endTime.hours}:${endTime.minutes}`
	};


	return _addTimeEntry(timeEntryInput, addTimeEntry);
};

/**
 * Given and array of times *(entry, lunch_out, lunch_in, exit), calculate the total laboured hours
 * @param {Object[]} storedTimes is an array of obj {hours, minutes}
 * @return {string} the total duration as HH:mm
 */
export const calculateLabouredHours = (storedTimes) => {
	const startTime = storedTimes[storedTimesIndex.startTime];
	const startBreakTime = storedTimes[storedTimesIndex.startBreakTime];
	const endBreakTime = storedTimes[storedTimesIndex.endBreakTime];
	const endTime = storedTimes[storedTimesIndex.endTime];

	const labouredHoursOnDay = new TimeDuration();
	labouredHoursOnDay.add(endTime);
	labouredHoursOnDay.subtract(endBreakTime);
	labouredHoursOnDay.add(startBreakTime);
	labouredHoursOnDay.subtract(startTime);

	return labouredHoursOnDay.toString();
};

const getContractedHoursUpToDate = (contractedHours, businessDaysUpToDate) =>
	new TimeDuration(contractedHours).multiplyBy(businessDaysUpToDate);

const getLabouredHoursUpToDate = (controlDate, timeEntries, labouredHoursOnDay = false) => {
	const timeEntriesUpToDate = timeEntries
		.filter(dayEntry => moment(dayEntry.date).isSameOrBefore(controlDate, 'day'));
	const minutesUpToDateWithActiveDay = timeEntriesUpToDate
		.map((dayEntry) => {
			const hasActiveDay = moment(dayEntry.date).isSame(controlDate, 'day') &&
				labouredHoursOnDay;
			return new TimeDuration(hasActiveDay ?
				labouredHoursOnDay : dayEntry.total).toMinutes();
		});

	const minutesLabouredUpToDate = minutesUpToDateWithActiveDay
		.reduce((totalUpToNow, dayTotal) => totalUpToNow + dayTotal, 0);

	return new TimeDuration(minutesLabouredUpToDate);
};

export const calculateHoursBalanceUpToDate = (controlDate, params) => {
	const {
		labouredHoursOnDay,
		contractedHoursForADay,
		timeEntries
	} = params;

	const businessDaysUpToDate = controlDate.day() > 5 ? 5 : controlDate.day();

	const contractedHoursUpToDate = getContractedHoursUpToDate(
		contractedHoursForADay,
		businessDaysUpToDate
	);

	const labouredHoursUpToDate = getLabouredHoursUpToDate(
		controlDate,
		timeEntries,
		labouredHoursOnDay
	);

	return {
		contractedHoursUpToDate,
		labouredHoursUpToDate
	};
};

/**
 * Given and array of times, check if it's completed and if hours are increasing
 * @param {Object[]} times is an array of obj {hours, minutes}
 * @return {bool} if it is a valid array
 */
export const timesAreValid = (times) => {
	let comparisonTerm = 0;
	const isSequentialTime = (time) => {
		if (time && timeIsValid(time)) {
			const date = new Date(2017, 0, 1, time.hours, time.minutes, 0, 0);
			const isLaterThanComparison = date > comparisonTerm;
			comparisonTerm = Number(date);
			return isLaterThanComparison;
		}
		return false;
	};

	return times.every(isSequentialTime);
};

export const dismemberTimeString = (timeString) => {
	const [hours, minutesRaw] = String(timeString).split(':');
	const numberMinutes = parseInt(minutesRaw, 10);
	const minutes = (numberMinutes < 10) ? `0${numberMinutes}` : minutesRaw;
	return { hours, minutes };
};

export const isDayBlockedInPast = (day) => {
	const today = moment();
	if (day.isSameOrAfter(today, 'day')) {
		return false;
	}
	const MONDAY = 1;
	const todayIsMonday = today.day() === MONDAY;
	// if it is monday then last week is still valid
	if (todayIsMonday) {
		const lastFriday = today.subtract(3, 'days');
		return day.isBefore(lastFriday, 'week');
	}
	return day.isBefore(today, 'week');
};
