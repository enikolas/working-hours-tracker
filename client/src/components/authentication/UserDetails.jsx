import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { graphql } from '@apollo/client/react/hoc';
import { compose } from 'recompose';

import * as queries from '../../queries';
import strings from '../../utils/strings';
import { getAuthToken, removeAuthToken } from './token';
import { clearTodayStorage } from '../../utils';
import UserOptions from '../userOptions/UserOptions';

import './UserDetails.css';

const _logout = (event) => {
	event.preventDefault();
	removeAuthToken();
	clearTodayStorage();
	window.location.reload();
};

class UserDetails extends Component {
	constructor(props) {
		super(props);

		this.state = {
			authenticated: false
		};

		this.openMonthlyReport = this.openMonthlyReport.bind(this);
		this.changePassword = this.changePassword.bind(this);
	}

	componentWillReceiveProps(nextProps) {
		const { userDetails, loading, error } = nextProps.userDetailsQuery;

		if (error || (!loading && !userDetails)) {
			removeAuthToken();
			this.setState({ authenticated: false });
		} else {
			const token = getAuthToken();
			const authenticated = Boolean(token && userDetails.name);
			this.setState({ authenticated });
		}
	}

	/* eslint-disable class-methods-use-this */

	openMonthlyReport(monthData) {
		console.log({ monthData });
	}

	async changePassword(passwordData) {
		const currentPassword = passwordData.old;
		const newPassword = passwordData.new;

		const response = await this.props.changePassword({
			variables: {
				currentPassword,
				newPassword
			}
		});

		if (response) {
			if (response.errors && response.errors.length) {
				// TODO: shows error message
			} else {
				removeAuthToken();
				window.location.reload();
			}
		}
	}

	/* eslint-enable class-methods-use-this */

	render() {
		const { userDetails, loading, error } = this.props.userDetailsQuery;

		if (loading) {
			return <div>Loading...</div>;
		}

		const { authenticated } = this.state;

		if (authenticated && !error) {
			const [firstName] = userDetails.name.split(' ');

			return (
				<div className="user-details">
					<div className="employee-name">
						{`${strings.helloName} ${firstName}`}
					</div>
					<nav>
						<UserOptions
							reportCall={this.openMonthlyReport}
							passwordCall={this.changePassword}
						/>
						<li>
							<button className="logout" onClick={_logout}>
								{strings.logout}
							</button>
						</li>
					</nav>
				</div>
			);
		}
		return null;
	}
}

export default compose(
	graphql(queries.userDetails, { name: 'userDetailsQuery' }),
	graphql(queries.changePassword, { name: 'changePassword' })
)(UserDetails);

UserDetails.propTypes = {
	userDetailsQuery: PropTypes.object,
	changePassword: PropTypes.func
};

UserDetails.defaultProps = {
	userDetailsQuery: {},
	changePassword: () => {}
};
