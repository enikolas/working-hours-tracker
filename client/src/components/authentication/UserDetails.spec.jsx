import React from 'react';
import { shallow } from 'enzyme';

import UserDetails from './UserDetails';

import * as utils from '../../utils';
import * as token from './token';

const defaultUserDetailsQuery = {
	loading: false,
	userDetails: {
		name: 'JOHN DOE',
		dailyContractedHours: '8:00',
		lastFridayBalance: ' 2:20',
		__typename: 'UserDetails'
	},
	error: false
};

describe('UserDetails', () => {
	describe('render', () => {
		it('should render loading if loading', () => {
			const wrapper = shallow(<UserDetails
				userDetailsQuery={{ ...defaultUserDetailsQuery, loading: true }}
			/>);
			expect(wrapper).toMatchSnapshot();
		});
		it('should render a div for login if not auth', () => {
			const wrapper = shallow(<UserDetails userDetailsQuery={defaultUserDetailsQuery} />);
			expect(wrapper).toMatchSnapshot();
		});
		it('should render a div for login if auth but has error', () => {
			const wrapper = shallow(<UserDetails userDetailsQuery={defaultUserDetailsQuery} />);
			wrapper.setProps({ userDetailsQuery: { ...defaultUserDetailsQuery, error: true } });
			wrapper.setState({ authenticated: true });
			expect(wrapper).toMatchSnapshot();
		});
		it('should render a welcome div if user auth and no errors', () => {
			const wrapper = shallow(<UserDetails userDetailsQuery={defaultUserDetailsQuery} />);
			wrapper.setState({ authenticated: true });
			expect(wrapper).toMatchSnapshot();
		});
	});
	describe('logout', () => {
		it('should be called at logout button', () => {
			utils.clearTodayStorage = jest.fn();		// Will remove infos from user of localStorage
			token.removeAuthToken = jest.fn();			//	^
			window.location.reload = jest.fn();			// Reload window to logout

			const wrapper = shallow(<UserDetails userDetailsQuery={defaultUserDetailsQuery} />);
			wrapper.setState({ authenticated: true });
			const logOutButton = wrapper.find('button');
			logOutButton.simulate('click', { preventDefault: jest.fn() });

			expect(utils.clearTodayStorage).toHaveBeenCalled();
			expect(token.removeAuthToken).toHaveBeenCalled();
			expect(window.location.reload).toHaveBeenCalled();
		});
	});

	describe('componentWillReceiveProps', () => {
		it('should authenticate if receive props', () => {
			token.getAuthToken = jest.fn(() => ('superSecureToken'));

			const loadingUserDetails = { ...defaultUserDetailsQuery, loading: true };
			const wrapper = shallow(<UserDetails userDetailsQuery={loadingUserDetails} />);

			wrapper.setProps({ userDetailsQuery: defaultUserDetailsQuery });

			expect(wrapper.state('authenticated')).toBeTruthy();
		});
	});
});
