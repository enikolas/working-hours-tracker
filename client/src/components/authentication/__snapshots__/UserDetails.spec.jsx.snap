// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDetails render should render a div for login if auth but has error 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UserDetails
    changePassword={[Function]}
    userDetailsQuery={
      Object {
        "error": true,
        "loading": false,
        "userDetails": Object {
          "__typename": "UserDetails",
          "dailyContractedHours": "8:00",
          "lastFridayBalance": " 2:20",
          "name": "JOHN DOE",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`UserDetails render should render a div for login if not auth 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UserDetails
    changePassword={[Function]}
    userDetailsQuery={
      Object {
        "error": false,
        "loading": false,
        "userDetails": Object {
          "__typename": "UserDetails",
          "dailyContractedHours": "8:00",
          "lastFridayBalance": " 2:20",
          "name": "JOHN DOE",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`UserDetails render should render a welcome div if user auth and no errors 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UserDetails
    changePassword={[Function]}
    userDetailsQuery={
      Object {
        "error": false,
        "loading": false,
        "userDetails": Object {
          "__typename": "UserDetails",
          "dailyContractedHours": "8:00",
          "lastFridayBalance": " 2:20",
          "name": "JOHN DOE",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="employee-name"
        >
          Olá  JOHN
        </div>,
        <nav>
          <UserOptions
            passwordCall={[Function]}
            reportCall={[Function]}
          />
          <li>
            <button
              className="logout"
              onClick={[Function]}
            >
              Sair
            </button>
          </li>
        </nav>,
      ],
      "className": "user-details",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Olá  JOHN",
          "className": "employee-name",
        },
        "ref": null,
        "rendered": "Olá  JOHN",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <UserOptions
              passwordCall={[Function]}
              reportCall={[Function]}
            />,
            <li>
              <button
                className="logout"
                onClick={[Function]}
              >
                Sair
              </button>
            </li>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "passwordCall": [Function],
              "reportCall": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <button
                className="logout"
                onClick={[Function]}
              >
                Sair
              </button>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Sair",
                "className": "logout",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Sair",
              "type": "button",
            },
            "type": "li",
          },
        ],
        "type": "nav",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="employee-name"
          >
            Olá  JOHN
          </div>,
          <nav>
            <UserOptions
              passwordCall={[Function]}
              reportCall={[Function]}
            />
            <li>
              <button
                className="logout"
                onClick={[Function]}
              >
                Sair
              </button>
            </li>
          </nav>,
        ],
        "className": "user-details",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Olá  JOHN",
            "className": "employee-name",
          },
          "ref": null,
          "rendered": "Olá  JOHN",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <UserOptions
                passwordCall={[Function]}
                reportCall={[Function]}
              />,
              <li>
                <button
                  className="logout"
                  onClick={[Function]}
                >
                  Sair
                </button>
              </li>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "passwordCall": [Function],
                "reportCall": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <button
                  className="logout"
                  onClick={[Function]}
                >
                  Sair
                </button>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Sair",
                  "className": "logout",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Sair",
                "type": "button",
              },
              "type": "li",
            },
          ],
          "type": "nav",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`UserDetails render should render loading if loading 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UserDetails
    changePassword={[Function]}
    userDetailsQuery={
      Object {
        "error": false,
        "loading": true,
        "userDetails": Object {
          "__typename": "UserDetails",
          "dailyContractedHours": "8:00",
          "lastFridayBalance": " 2:20",
          "name": "JOHN DOE",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "Loading...",
    },
    "ref": null,
    "rendered": "Loading...",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Loading...",
      },
      "ref": null,
      "rendered": "Loading...",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
